sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -48,6 +50,63 @@ def csr_row_norms(X):
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
+def sparse_mean_variance(X):
+    """Compute the variance of a sparse matrix.
+    Will copy data to an array of doubles if input data
+    are not np.float32 or np.float64 types.
+
+    Parameters
+    ----------
+    X: Scipy sparse matrix, shape (n_samples, n_features)
+        Input data.
+
+    Returns
+    -------
+
+    mean: float scalar
+        Mean of array
+
+    variance: float scalar
+        Variance of array
+    """
+    cdef:
+        double sum = 0
+        double sum_squares = 0
+        double variance = 0
+        double n_elements = X.shape[0] * X.shape[1]
+        np.ndarray[DOUBLE, ndim=1, mode="c"] X_data_d
+        np.ndarray[FLOAT, ndim=1, mode="c"] X_data_f
+        int len_data
+        double datum
+    len_data = len(X.data)
+
+    X_data = X.data
+    if X_data.dtype not in (np.float64, np.float32):
+        warnings.warn("Copying sparse array data to array of "
+                      "doubles for computation.")
+        X_data = np.asarray(X_data, dtype=np.float64)  # Copy!
+    if X_data.dtype == np.float64:
+        X_data_d = X_data
+        for i_datum in xrange(len_data):
+            datum = X_data_d[i_datum]
+            sum += datum
+            sum_squares += datum * datum
+    elif X_data.dtype == np.float32:
+        X_data_f = X_data
+        for i_datum in xrange(len_data):
+            datum = X_data_f[i_datum]
+            sum += datum
+            sum_squares += datum * datum
```

Comment:
I think this function should be rewritten to leverage Cython fused types:

```cython
from cython cimport floating
from numpy import float16_t, float32_t, float64_t, ndarray

ctypedef fused myfloating:
    float16_t
    float32_t
    float64_t

def sparse_mean_variance(X):
    # check data type and make a copy with astype if necessary
    # then call _floating_mean_variance

cdef _floating_mean_variance(ndarray[floating, ndim=2] array):
    # put the actual computation here

```

This would also allow to avoid a memory copy with float16 data. To check the case where a memory copy is required, just use int32 or int64 typed input.